### PR TITLE
CMake build info wrapper method for "magic lines"

### LINKED ISSUE #2590: Provide a tool to inject the conan cmake "magic" lines
Feature request:

To simplify the recipes with the classical `replace_in_file`, we could provide a helper, could be inside the `CMake` helper as a static method (?) that searches the `project()` in a `CMakeLists.txt`and inject the lines:
```
include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
CONAN_BASIC_SETUP()
```

Proposed: `CMake.insert_conan_build_info(folder)` 
Maybe it could detect the `install_folder` to adjust the install dir if present?

### PR DESCRIPTION
**Pull request by danimtb**
- [x] Refer to the issue that supports this Pull Request: closes #2590
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. Also adding a description of the changes in the ``changelog.rst`` file. https://github.com/conan-io/docs

With the bincrafters "wrapper" approach. It seems cleaner and easier to me but you have to configure your sources in a subdirectory.

### CODE REVIEW

**Comment by lasote** [comment]
> With the bincrafters "wrapper" approach. It seems cleaner and easier for me but you have to configure your sources in a subdirectory.

Maybe you are right, but if we want to generate a `CMakeLists.txt` file to include the project one, the implemented feature is not ok:

1. The name `insert_conan_build_info` makes no sense. `write_base_cmakelist` or something similar would be better.
2. The parameter should be a path to the original CMakeLists.txt
3. It should be check if the cmakelists is there and that it is not in the same directory than we are going to generate ours.
4. Probably it should not be an static method and should check the conanfile generators to make sure that the conanbuildinfo.cmake is in the `self.build_folder`.
5. Even with the previous points, I'm not sure if I prefer to generate the file by hand (explicit)

**Comment by lasote** [comment]
6. Given that we have a pending deep task of restructuring the build_helpers/generators, I won't add more features if they are not completely necessary. I vote to not include this in 1.5

**Comment by danimtb** [comment]
I understan the points and the only thing to push this forward is to make recipes cleaner. But I see restructuration of build_helpers/generators is more relevant.

Feel free to close or clear the milestone. Thanks!

**Comment by lasote** [comment]
Let's wait for @memsharded opinion

File: conans/test/integration/cmake_magic_lines_test.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+"""
+
+cmake = """
+cmake_minimum_required(VERSION 2.8)
+PROJECT(MyHello)
+ADD_LIBRARY(hello hello.cpp)
+"""
+
+
+@attr("slow")
```
Comment (memsharded) [inline conans/test/integration/cmake_magic_lines_test.py]: Not another slow test for a "replace" functionality. Just check the files are there.

File: conans/client/build/cmake.py [hunk 2]
```diff
@@ -526,3 +526,11 @@ def _get_cpp_standard_vars(self):
         ret["CONAN_STD_CXX_FLAG"] = cppstd_flag(self._compiler, self._compiler_version,
                                                 self._cppstd)
         return ret
+
+    @staticmethod
+    def insert_conan_build_info(source_subfolder):
+        wrapper = """project(cmake_wrapper)
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
```
Comment (memsharded) [inline conans/client/build/cmake.py]: It seems that some basic parameterization is needed: TARGETS, OUTPUTDIRS...

**Comment by lasote** [comment]
As we are still investigating if another way rather than injecting these lines is possible I think we can close this one until it is clarified. Not a big benefit.

**Pull request closed without merging**